src/liballoc/collections/btree/set.rs

```diff
@@ -1365,30 +1406,20 @@ impl<T: Ord> FusedIterator for Intersection<'_, T> {}
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> Clone for Union<'_, T> {
     fn clone(&self) -> Self {
-        Union {
-            a: self.a.clone(),
-            b: self.b.clone(),
-        }
+        Union(self.0.clone())
     }
 }
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<'a, T: Ord> Iterator for Union<'a, T> {
     type Item = &'a T;
 
     fn next(&mut self) -> Option<&'a T> {
-        match cmp_opt(self.a.peek(), self.b.peek(), Greater, Less) {
-            Less => self.a.next(),
-            Equal => {
-                self.b.next();
-                self.a.next()
-            }
-            Greater => self.b.next(),
-        }
+        let (a_next, b_next) = self.0.next();
+        a_next.or(b_next)
     }
 
     fn size_hint(&self) -> (usize, Option<usize>) {
-        let a_len = self.a.len();
-        let b_len = self.b.len();
+        let (a_len, b_len) = (self.0.a.len(), self.0.b.len());
         (max(a_len, b_len), Some(a_len + b_len))
```

Comment:
(Preexisting) The length sum should be `a_len.checked_add(b_len)` and overflow to None.

Reply:
Ok with comment explaining max set length is limited.

Reply:
Well spotted. I certainly don't know what's been argued, but I'd argue myself that even when usize is 16 bit (surely someone ported Rust for Commodore 64), you cannot create a set with >= 32k elements. If each element takes up 1 byte, there are only 256 of them anyways. If they take up 2 bytes, a set will have to stay shy of 32k elements. So even if you ask for the union of a 32k set with itself, size_hint should be safe.

Reply:
I believe each element could be 0 size (and still none equal to any other), but internal overhead will create equivalent restrictions.